.github/workflows/build-test-distribute.yaml

```diff
@@ -170,3 +173,161 @@ jobs:
           trap on_exit EXIT
           make docker/push
           make docker/manifest
+  gen_e2e_matrix:
+    runs-on: ubuntu-latest
+    needs: ["build"]
+    if: ${{ !contains(github.event.pull_request.labels.*.name, 'ci/skip-test') && !contains(github.event.pull_request.labels.*.name, 'ci/skip-e2e-test') }}
+    outputs:
+      slow: ${{ steps.generate-matrix.outputs.slow }}
+      main: ${{ steps.generate-matrix.outputs.main }}
+      legacy_kds: ${{ steps.generate-matrix.outputs.legacy_kds }}
+      cni: ${{ steps.generate-matrix.outputs.cni }}
+    steps:
+      - name: Install jq
+        run: |
+          sudo apt-get install -y jq
+      - id: generate-matrix
+        name: Generate matrix
+        run: |
+          BASE_MATRIX=$(cat <<-END
+          {
+            "target": ["kubernetes", "universal", "multizone"],
+            "k8sVersion": ["kind", "kindIpv6", "${{ env.K8S_MIN_VERSION }}", "${{ env.K8S_MAX_VERSION }}"],
+            "arch": ["amd64", "arm64"],
+            "cniNetworkPlugin": ["flannel"],
+            "legacyKDS": [false],
+            "exclude":[
+              {"target": "kubernetes", "k8sVersion":"kind"},
+              {"target": "multizone", "k8sVersion":"kind"},
+              {"target":"universal", "k8sVersion":"${{ env.K8S_MIN_VERSION }}"}, 
+              {"target":"universal", "k8sVersion":"${{ env.K8S_MAX_VERSION }}"}
+            ]
+          }
+          END
+          )
+          
+          SLOW=$(echo $BASE_MATRIX | jq -rc '.target = [""]')
+          MAIN=$(echo $BASE_MATRIX | jq -rc)
+          
+          SINGLE_ENV='{"k8sVersion": ["${{ env.K8S_MAX_VERSION }}"], "target": ["multizone"], "arch":["amd64"]}'
+          LEGACY_KDS=$(echo '{}' | jq -rc ".legacyKDS = [true] | . += $SINGLE_ENV")
+          CNI=$(echo '{}' | jq -rc ".cniNetworkPlugin = [\"calico\"] | . += $SINGLE_ENV")
+
+          RUN_FULL_MATRIX="${{ github.event_name == 'push' || contains(github.event.pull_request.labels.*.name, 'ci/run-full-matrix') }}"
+          if [[ "$RUN_FULL_MATRIX" == "false" ]]; then
+            echo "Skipping non priority e2e tests, because no label 'ci/run-full-matrix' was found on the pull request and not pushing on a releasable branch."
+
+            SLOW=''         # target==""
+            LEGACY_KDS=''   # legacyKDS=="true"
+            CNI=''          # cniNetworkPlugin=="calico"
+          
+            # non priority main e2e
+            EXCLUDE_ARM64='[{"arch": "arm64"}]'
+            EXCLUDE_MAIN_NON_PRIORITY='[{"k8sVersion":"kindIpv6"}, {"k8sVersion": "${{ env.K8S_MIN_VERSION }}"}]'
+            MAIN=$(echo $MAIN | jq -rc ".exclude += $EXCLUDE_ARM64 | .exclude += $EXCLUDE_MAIN_NON_PRIORITY")
+          fi
+
+          echo "slow=$SLOW" >> $GITHUB_OUTPUT
+          echo "main=$MAIN" >> $GITHUB_OUTPUT
+          echo "legacy_kds=$LEGACY_KDS" >> $GITHUB_OUTPUT
+          echo "cni=$CNI" >> $GITHUB_OUTPUT
```

Comment:
only 2 outputs to this job:

test_e2e  is `slow`
test_e2e_env is a merge of `main`, `cni` and `legacy_kds` do not hesitate to use `include` on your matrix to make this easier.

Also has the 2 matrices (test_e2e and test_e2e_env) are very different maybe just have 2 `BASE_MATRIX`.

Reply:
Other option have 4 different matrices which map make targets:

- test/e2e
- test/e2e-universal
- test/e2e-multizone
- test/e2e-kubernetes

Then each have their own matrix.

Reply:
Updated code with the first approach.